File: patterns/eventsourcing/EventSourcing.UnitTests/BookingTests.cs
using Bogus;
using EventSourcing.Domain.Core;

namespace EventSourcing.UnitTests;

public class BookingTests
{
    private readonly IReadOnlyList<Attendee> _attendees =
        [NewAttendee()];

    [Fact]
    public void CreateNewBooking_Append_BookingCreated()
    {
        // arrange
        var room = MeetingRoom.Helsinki;
        var startTime = DateTime.Now;
        var endTime = startTime.AddHours(1);

        // act
        Booking booking = new Booking(Guid.NewGuid(), room, startTime, endTime, _attendees);

        // assert
        Assert.Equal(room, booking.MeetingRoom);
        Assert.Equal(startTime, booking.StartTime);
        Assert.Equal(endTime, booking.EndTime);
        Assert.Equal(_attendees, booking.Attendees);
        Assert.Single(booking.Events);
        Assert.IsType<BookingCreated>(booking.Events[0]);
    }

    [Fact]
    public void UpdateTimeSlot_Append_BookedTimeSlotChanged()
    {
        // arrange
        var room = MeetingRoom.Helsinki;
        var now = DateTime.Now;
        Booking booking = new Booking(Guid.NewGuid(), room, now, now.AddHours(1), _attendees);

        // act
        var newStartTime = now.AddHours(1);
        var newEndTime = now.AddHours(2);
        booking.UpdateTimeSlot(newStartTime, newEndTime);

        // assert
        Assert.Equal(newStartTime, booking.StartTime);
        Assert.Equal(newEndTime, booking.EndTime);

        Assert.Equal(2, booking.Events.Count);
        Assert.IsType<BookingCreated>(booking.Events[0]);
        Assert.IsType<BookedTimeSlotChanged>(booking.Events[1]);
    }

    [Fact]
    public void UpdateMeetingRoom_Append_MeetingRoomChanged()
    {
        // arrange
        var startTime = DateTime.Now;
        var endTime = startTime.AddHours(1);
        Booking booking = new Booking(Guid.NewGuid(), MeetingRoom.Helsinki, startTime, endTime, _attendees);

        // act
        var newMeetingRoom = MeetingRoom.Paris;
        booking.UpdateMeetingRoom(newMeetingRoom);

        // assert
        Assert.Equal(newMeetingRoom, booking.MeetingRoom);

        Assert.Equal(2, booking.Events.Count);
        Assert.IsType<BookingCreated>(booking.Events[0]);
        Assert.IsType<MeetingRoomChanged>(booking.Events[1]);
    }

    [Fact]
    public void AddNewAttendee_Append_MeetingRoomChanged()
    {
        // arrange
        var startTime = DateTime.Now;
        var endTime = startTime.AddHours(1);
        Booking booking = new Booking(Guid.NewGuid(), MeetingRoom.Helsinki, startTime, endTime, _attendees);

        // act
        var newAttendee = NewAttendee();
        booking.AddAttendee(newAttendee);

        // assert
        Assert.Equal([.._attendees, newAttendee], booking.Attendees);

        Assert.Equal(2, booking.Events.Count);
        Assert.IsType<BookingCreated>(booking.Events[0]);
        Assert.IsType<AttendeeAdded>(booking.Events[1]);
    }

    [Fact]
    public void BookingSubsequentUpdates()
    {
        // arrange
        var now = DateTime.Now;
        Booking booking = new Booking(Guid.NewGuid(), MeetingRoom.Helsinki, now, now.AddHours(1), _attendees);

        // act
        var newRoom = MeetingRoom.Paris;
        booking.UpdateMeetingRoom(newRoom);

        // act
        var newStartTime = now.AddHours(1);
        var newEndTime = now.AddHours(2);
        booking.UpdateTimeSlot(newStartTime, newEndTime);

        // assert
        Assert.Equal(newRoom, booking.MeetingRoom);
        Assert.Equal(newStartTime, booking.StartTime);
        Assert.Equal(newEndTime, booking.EndTime);

        Assert.Equal(3, booking.Events.Count);
        Assert.IsType<BookingCreated>(booking.Events[0]);
        Assert.IsType<MeetingRoomChanged>(booking.Events[1]);
        Assert.IsType<BookedTimeSlotChanged>(booking.Events[2]);
    }

    private static Attendee NewAttendee() =>
        new(new Faker().Person.FirstName, new Faker().Person.LastName, new Faker().Person.Email);
}

File: patterns/eventsourcing/fsharp/EventSourcing.UnitTests/BookingTests.fs
// ReSharper disable FSharpRedundantUnionCaseFieldPatterns

module EventSourcing.UnitTests.BookingTests

open System

open Bogus
open Xunit

open EventSourcing.Domain.Enums.MeetingRoom
open EventSourcing.Domain.Entities.Booking

let NewAttendee () : Attendee =
    let faker = Faker()

    { FirstName = faker.Person.FirstName
      LastName = faker.Person.LastName
      Email = faker.Internet.Email() }


[<Fact>]
let ``Create new booking appends BookingCreated event`` () =
    // arrange
    let id = Guid.CreateVersion7()
    let room = MeetingRoom.Helsinki
    let startTime = DateTime.Now
    let endTime = startTime.AddHours(1)
    let attendees = [ NewAttendee() ]

    // act
    let booking = Booking.Create(id, room, startTime, endTime, attendees)

    // assert
    Assert.Equal(room, booking.MeetingRoom)
    Assert.Equal(startTime, booking.StartTime)
    Assert.Equal(endTime, booking.EndTime)
    Assert.Equal<Attendee>(attendees, booking.Attendees)

    let events = booking.Events
    Assert.Single(events) |> ignore

    match events[0] with
    | BookingCreated(actualId, actualMeetingRoom, actualStartTime, actualEndTime, actualAttendees) ->
        Assert.Equal(id, actualId)
        Assert.Equal(room, actualMeetingRoom)
        Assert.Equal(startTime, actualStartTime)
        Assert.Equal(endTime, actualEndTime)
        Assert.Equal<Attendee>(attendees, actualAttendees)
    | _ -> Assert.Fail("The event is not a BookingCreated event")

[<Fact>]
let ``Update time slot appends BookedTimeSlotChanged event`` () =
    // arrange
    let initialStartTime = DateTime.Now
    let initialEndTime = initialStartTime.AddHours(1)
    let attendees = [ NewAttendee() ]

    let booking =
        Booking.Create(Guid.CreateVersion7(), MeetingRoom.Helsinki, initialStartTime, initialEndTime, attendees)

    // act
    let newStartTime = initialStartTime.AddHours(1)
    let newEndTime = newStartTime.AddHours(2)
    let booking = booking.UpdateTimeSlot(newStartTime, newEndTime)

    // assert
    Assert.Equal(booking.StartTime, booking.StartTime)
    Assert.Equal(booking.EndTime, booking.EndTime)

    let events = booking.Events
    Assert.Equal(2, events.Length)

    match events[0] with
    | BookingCreated(_, _, _, _, _) -> ()
    | _ -> Assert.Fail("The first event is not a BookingCreated event")

    match events[1] with
    | BookedTimeSlotChanged(actualStartTime, actualEndTime) ->
        Assert.Equal(newStartTime, actualStartTime)
        Assert.Equal(newEndTime, actualEndTime)
    | _ -> Assert.Fail("The second event is not a BookedTimeSlotChanged event")

[<Fact>]
let ``Update meeting room appends MeetingRoomChanged event`` () =
    // arrange
    let now = DateTime.Now
    let initialRoom = MeetingRoom.Helsinki

    let booking =
        Booking.Create(Guid.CreateVersion7(), initialRoom, now, now.AddHours(1), [ NewAttendee() ])

    // act
    let newRoom = MeetingRoom.Paris
    let booking = booking.UpdateMeetingRoom newRoom

    // assert
    Assert.Equal(newRoom, booking.MeetingRoom)

    let events = booking.Events
    Assert.Equal(2, events.Length)

    match events[0] with
    | BookingCreated(_, _, _, _, _) -> ()
    | _ -> Assert.Fail("The first event is not a BookingCreated event")

    match events[1] with
    | MeetingRoomChanged(actualRoom) -> Assert.Equal(newRoom, actualRoom)
    | _ -> Assert.Fail("The second event is not a MeetingRoomChanged event")

[<Fact>]
let ``Add new attendee appends AttendeeAdded event`` () =
    // arrange
    let now = DateTime.Now
    let attendees = [ NewAttendee() ]

    let booking =
        Booking.Create(Guid.CreateVersion7(), MeetingRoom.Helsinki, now, now.AddHours(1), attendees)

    // act
    let newAttendee = NewAttendee()
    let booking = booking.AddAttendee newAttendee

    // assert
    Assert.Equal<Attendee>(attendees @ [ newAttendee ], booking.Attendees)

    let events = booking.Events
    Assert.Equal(2, events.Length)

    match events[0] with
    | BookingCreated(_, _, _, _, _) -> ()
    | _ -> Assert.Fail("The first event is not a BookingCreated event")

    match events[1] with
    | AttendeeAdded(actualAttendees) -> Assert.Equal<Attendee>(attendees @ [ newAttendee ], actualAttendees)
    | _ -> Assert.Fail("The second event is not a AttendeeAdded event")

[<Fact>]
let ``Remove an attendee appends AttendeeRemoved event`` () =
    // arrange
    let now = DateTime.Now
    let attendees = [ NewAttendee(); NewAttendee() ]

    let booking =
        Booking.Create(Guid.CreateVersion7(), MeetingRoom.Helsinki, now, now.AddHours(1), attendees)

    // act
    let booking = booking.RemoveAttendee(attendees[0])

    // assert
    let expectedAttendees = attendees |> List.filter (fun a -> a <> attendees[0])
    Assert.Equal<Attendee>(expectedAttendees, booking.Attendees)

    let events = booking.Events
    Assert.Equal(2, events.Length)

    match events[0] with
    | BookingCreated(_, _, _, _, _) -> ()
    | _ -> Assert.Fail("The first event is not a BookingCreated event")

    match events[1] with
    | AttendeeRemoved(actualAttendees) -> Assert.Equal<Attendee>(expectedAttendees, actualAttendees)
    | _ -> Assert.Fail("The second event is not a AttendeeRemoved event")

[<Fact>]
let ``Booking subsequent updates`` () =
    // arrange
    let now = DateTime.Now
    let initialAttendee = NewAttendee()

    let booking =
        Booking.Create(
            Guid.CreateVersion7(),
            MeetingRoom.Helsinki,
            now,
            now.AddHours(1),
            [ NewAttendee(); initialAttendee ]
        )

    // act
    let newRoom = MeetingRoom.Paris
    let booking = booking.UpdateMeetingRoom(newRoom)

    // act
    let newStartTime = now.AddHours(1)
    let newEndTime = now.AddHours(2)
    let booking = booking.UpdateTimeSlot(newStartTime, newEndTime)

    // act
    let booking = booking.RemoveAttendee(booking.Attendees.Head)

    // act
    let newAttendee = NewAttendee()
    let booking = booking.AddAttendee(newAttendee)

    // assert
    Assert.Equal(newRoom, booking.MeetingRoom)
    Assert.Equal(newStartTime, booking.StartTime)
    Assert.Equal(newEndTime, booking.EndTime)
    Assert.Equal<Attendee>([ initialAttendee; newAttendee ], booking.Attendees)

    let events = booking.Events
    Assert.Equal(5, events.Length)

    match events[0] with
    | BookingCreated(_, _, _, _, _) -> ()
    | _ -> Assert.Fail("The first event is not a BookingCreated event")

    match events[1] with
    | MeetingRoomChanged _ -> ()
    | _ -> Assert.Fail("The second event is not a MeetingRoomChanged event")

    match events[2] with
    | BookedTimeSlotChanged _ -> ()
    | _ -> Assert.Fail("The third event is not a BookedTimeSlotChanged event")

    match events[3] with
    | AttendeeRemoved _ -> ()
    | _ -> Assert.Fail("The second event is not a AttendeeRemoved event")

    match events[4] with
    | AttendeeAdded _ -> ()
    | _ -> Assert.Fail("The second event is not a AttendeeAdded event")
